react/features/base/video-transform/components/TransformAwareView.js

```diff
@@ -0,0 +1,286 @@
+            };
+
+            // The A and D points of the transformed View.
+            const transformedLayout = {
+                a: {
+                    x: transformedCenterPoint.x - (transformedSize.width / 2),
+                    y: transformedCenterPoint.y - (transformedSize.height / 2)
+                },
+                d: {
+                    x: transformedCenterPoint.x + (transformedSize.width / 2),
+                    y: transformedCenterPoint.y + (transformedSize.height / 2)
+                }
+            };
+
+            let _maxOffset = MAX_OFFSET;
+
+            if (newScaleUnlimited < oldTransform.scale) {
+                // this is a negative scale event so we dynamycally reduce the
+                // maxOffset to get the screen back to the center on
+                // downscaling
+
+                _maxOffset = Math.min(MAX_OFFSET, MAX_OFFSET * (newScale - 1));
+            }
+
+            // This is the yet-unlimited transform that is desired
+            // by the user. This will be limited in the next step to avoid
+            // extreme or unwanted transformations.
+            const desiredTransform = {
+                scale: newScaleUnlimited,
+                translateX: newTranslateX,
+                translateY: newTranslateY
+            };
+
+            // Correct move matrix if it goes out of the view
+            // too much (_maxOffset).
+            newTranslateX -= Math.max(
+                transformedLayout.a.x - originalLayout.a.x - _maxOffset, 0
+            );
+            newTranslateX += Math.max(
+                originalLayout.d.x - transformedLayout.d.x - _maxOffset, 0
+            );
+            newTranslateY -= Math.max(
+                transformedLayout.a.y - originalLayout.a.y - _maxOffset, 0
+            );
+            newTranslateY += Math.max(
+                originalLayout.d.y - transformedLayout.d.y - _maxOffset, 0
+            );
+
+            const newTransform = {
+                scale: newScale,
+                translateX: Math.round(newTranslateX),
+                translateY: Math.round(newTranslateY)
+            };
+
+            if (oldTransform.scale !== newTransform.scale
+                || oldTransform.translateX !== newTransform.translateX
+                || oldTransform.translateY !== newTransform.translateY) {
+                // The transform has been changed, need to re-render.
+                this.setState({
+                    transform: newTransform
+                });
+            }
+
+            if (desiredTransform.scale !== newTransform.scale
+                || desiredTransform.translateX !== newTransform.translateX
+                || desiredTransform.translateY !== newTransform.translateY) {
+                // The transform has been limited, so we need to send an update.
+                if (typeof onTransformUpdate === 'function') {
+                    onTransformUpdate(newTransform);
```

Comment:
It's possible that I'm missing some details here, but what if the gesture component would be aware of the scale boundaries and be able cap it there ? Regards the x/y translation we could go from the -100% to +100% maybe ? This component could do any extra adjustments if it's not possible to move the view in some direction, but it could keep that to itself ? It would be good to get rid of this back and forth translation update, but maybe it's not easy to do. I'll try to think about it and maybe we can talk about that tomorrow. If I remember correctly that was one of Lyubo's feedbacks at the initial discussion.

Reply:
Isn't the gesture component aware of the layout size ? What if we would feed the size of the video into it ? Would it be able to calculate all offsets and limitations ?